tempesta_fw/http_limits.c

```diff
@@ -716,262 +674,319 @@ enum {
+			goto block;
 		}
 	}
 
 	/* Check for chunk count here to account for possible fragmentation
 	 * in HTTP status line. The rationale for not making this one of FSM
 	 * states is the same as for the code block above.
 	 */
-	if (hchnk_cnt && FSM_HDR_STATE(req->frang_st)) {
-		if (req->chunk_cnt > hchnk_cnt) {
-			frang_limmsg("HTTP header chunk count", req->chunk_cnt,
-				     hchnk_cnt, &FRANG_ACC2CLI(ra)->addr);
-			spin_unlock(&ra->lock);
+	if (hchnk_cnt && (req->chunk_cnt > hchnk_cnt)) {
+		frang_limmsg("HTTP header chunk count", req->chunk_cnt,
+			     hchnk_cnt, &FRANG_ACC2CLI(ra)->addr);
+		goto block;
+	}
+
+	return frang_http_req_hdrs_check(ra, conn, data);
+block:
+	return TFW_BLOCK;
+}
+
+static int
+frang_http_req_incomplete_body_check(FrangAcc *ra, TfwConn *conn,
+				     TfwFsmData *data)
+{
+	TfwHttpReq *req = (TfwHttpReq *)data->req;
+	__FRANG_CFG_VAR(body_len, http_body_len);
+	__FRANG_CFG_VAR(body_timeout, clnt_body_timeout);
+	__FRANG_CFG_VAR(bchunk_cnt, http_bchunk_cnt);
+	struct sk_buff *skb = data->skb;
+
+	frang_dbg("check incomplete request body for client %s, acc=%p\n",
+		  &FRANG_ACC2CLI(ra)->addr, ra);
+
+	/* CLRF after headers was parsed, but the body didn't arrive yet. */
+	if (TFW_STR_EMPTY(&req->body))
+		return TFW_PASS;
+	/*
+	 * Ensure that HTTP request body is coming without delays.
+	 * The timeout is between chunks of the body, so reset
+	 * the start time after each check.
+	 */
+	if (body_timeout && req->tm_bchunk && (skb != req->body.skb)) {
+		unsigned long start = req->tm_bchunk;
+		unsigned long delta = body_timeout;
+
+		if (time_is_before_jiffies(start + delta)) {
+			frang_limmsg("client body timeout",
+				     jiffies - start, delta,
+				     &FRANG_ACC2CLI(ra)->addr);
+			goto block;
+		}
+	}
+
+	/* Limit number of chunks in request body */
+	if (bchunk_cnt && (req->chunk_cnt > bchunk_cnt)) {
+		frang_limmsg("HTTP body chunk count", req->chunk_cnt,
+			     bchunk_cnt, &FRANG_ACC2CLI(ra)->addr);
+		goto block;
+	}
+
+	if (body_len && (req->body.len > body_len)) {
+		frang_limmsg("HTTP body length", req->body.len,
+			     body_len, &FRANG_ACC2CLI(ra)->addr);
+		goto block;
+	}
+
+
+	return TFW_PASS;
+block:
+	return TFW_BLOCK;
+}
+
+static int
+frang_http_req_trailer_check(FrangAcc *ra, TfwConn *conn,
+			     TfwFsmData *data)
+{
+	int r = TFW_PASS;
+	TfwHttpMsg *hmreq = (TfwHttpMsg *)data->req;
+	const TfwStr *field, *end, *dup, *dup_end;
+
+	if (!test_bit(TFW_HTTP_B_CHUNKED_TRAILER, hmreq->flags))
+		return TFW_PASS;
+	/*
+	 * Don't use special settings for the trailer part, keep on
+	 * using body limits.
+	 */
+	r = frang_http_req_incomplete_body_check(ra, conn, data);
+	if (!r)
+		r = frang_http_req_hdrs_check(ra, conn, data);
+	if (r)
+		return r;
+
+	if (!tfw_http_parse_is_done(hmreq))
+		return TFW_POSTPONE;
+	/*
+	 * Block request if the same header appear in both main and
+	 * trailer headers part. Some intermediates doesn't read trailers, so
+	 * request processing may differ.
```

Comment:
Good improvement. I guess this problem with different processing of duplicate trailer headers is defined in some RFC - please add a reference to the RFC and chapter.

Reply:
not fixed